### PR TITLE
Separate DALI operators shared-object library

### PR DESCRIPTION
**Pull request by jantonguirao**
Depends on #1369 

#### Why we need this PR?
Need to separate DALI operators from DALI lib basics

#### What happened in this PR?
- Moved dali/pipeline/operators to dali/operators

**JIRA TASK**: [DALI-1089]

### CODE REVIEW

File: dali/operators/crop/crop.h [hunk 1]
```diff
@@ -22,10 +22,10 @@
 #include "dali/core/error_handling.h"
 #include "dali/kernels/scratch.h"
 #include "dali/kernels/tensor_shape.h"
-#include "dali/pipeline/operators/common.h"
-#include "dali/pipeline/operators/crop/crop_attr.h"
-#include "dali/pipeline/operators/crop/slice_base.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/common.h"
```
Comment (klecki) [inline dali/operators/crop/crop.h]: Can you sort the includes in such cases?

File: dali/operators/crop/slice.h [hunk 2]
```diff
@@ -12,16 +12,16 @@
...
 #include <utility>
 #include <vector>
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
-#include "dali/pipeline/operators/common.h"
-#include "dali/pipeline/operators/crop/slice_base.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/operators/crop/slice_base.h"
+#include "dali/pipeline/operator/operator.h"
```
Comment (klecki) [inline dali/operators/crop/slice.h]: And here

File: dali/operators/crop/slice_attr.h [hunk 3]
```diff
@@ -12,15 +12,15 @@
...
 #include <utility>
 #include <vector>
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
-#include "dali/pipeline/operators/common.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/pipeline/operator/operator.h"
 #include "dali/util/crop_window.h"
 #include "dali/core/tensor_layout.h"
```
Comment (klecki) [inline dali/operators/crop/slice_attr.h]: Maybe even here?

File: dali/operators/crop/slice_base.h [hunk 4]
```diff
@@ -12,18 +12,18 @@
...
 #include <tuple>
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
 #include "dali/kernels/scratch.h"
-#include "dali/pipeline/operators/common.h"
-#include "dali/pipeline/operators/crop/crop_attr.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/operators/crop/crop_attr.h"
+#include "dali/pipeline/operator/operator.h"
```
Comment (klecki) [inline dali/operators/crop/slice_base.h]: And here

File: dali/operators/decoder/cache/image_cache_blob.cc [hunk 5]
```diff
@@ -19,7 +19,7 @@
 #include "dali/kernels/alloc.h"
 #include "dali/core/span.h"
 #include "dali/pipeline/data/backend.h"
-#include "dali/pipeline/operators/decoder/cache/image_cache_blob.h"
+#include "dali/operators/decoder/cache/image_cache_blob.h"
```
Comment (klecki) [inline dali/operators/decoder/cache/image_cache_blob.cc]: And here

File: dali/operators/decoder/image_decoder.cc [hunk 6]
```diff
@@ -12,8 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/decoder/host/host_decoder.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/decoder/host/host_decoder.h"
```
Comment (klecki) [inline dali/operators/decoder/image_decoder.cc]: And here

File: dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_random_crop.cc [hunk 7]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "dali/pipeline/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_random_crop.h"
+#include "dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_random_crop.h"
```
Comment (klecki) [inline dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_random_crop.cc]: Btw, is there a reason why this is included above system header?

File: dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_slice.cc [hunk 8]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "dali/pipeline/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_slice.h"
+#include "dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_slice.h"
```
Comment (klecki) [inline dali/operators/decoder/nvjpeg/decoupled_api/fused/nvjpeg_decoder_cpu_slice.cc]: And here

File: dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_gpu.h [hunk 9]
```diff
@@ -12,15 +12,15 @@
...
+#define DALI_OPERATORS_DECODER_NVJPEG_DECOUPLED_API_NVJPEG_DECODER_GPU_H_
 
 #include <functional>
 #include <utility>
 #include <vector>
 
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/decoder/nvjpeg/decoupled_api/nvjpeg_helper.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_helper.h"
```
Comment (klecki) [inline dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_gpu.h]: And here

File: dali/operators/displacement/jitter.h [hunk 10]
```diff
@@ -13,15 +13,15 @@
...
 
 #include <ctgmath>
 #include <vector>
 #include "dali/core/host_dev.h"
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/displacement/displacement_filter.h"
-#include "dali/pipeline/operators/util/randomizer.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/displacement/displacement_filter.h"
+#include "dali/operators/util/randomizer.h"
```
Comment (klecki) [inline dali/operators/displacement/jitter.h]: And here

File: dali/operators/displacement/rotate.h [hunk 11]
```diff
@@ -13,13 +13,13 @@
...
-#define DALI_PIPELINE_OPERATORS_DISPLACEMENT_ROTATE_H_
+#ifndef DALI_OPERATORS_DISPLACEMENT_ROTATE_H_
+#define DALI_OPERATORS_DISPLACEMENT_ROTATE_H_
 
 #include <vector>
 #include <cmath>
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/displacement/warpaffine.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/displacement/warpaffine.h"
```
Comment (klecki) [inline dali/operators/displacement/rotate.h]: And here

File: dali/operators/displacement/sphere.h [hunk 12]
```diff
@@ -13,13 +13,13 @@
...
-#define DALI_PIPELINE_OPERATORS_DISPLACEMENT_SPHERE_H_
+#ifndef DALI_OPERATORS_DISPLACEMENT_SPHERE_H_
+#define DALI_OPERATORS_DISPLACEMENT_SPHERE_H_
 
 #include <ctgmath>
 #include <vector>
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/displacement/displacement_filter.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/displacement/displacement_filter.h"
```
Comment (klecki) [inline dali/operators/displacement/sphere.h]: And here

File: dali/operators/displacement/warp.h [hunk 13]
```diff
@@ -28,8 +28,8 @@
 #include "dali/kernels/kernel_manager.h"
 #include "dali/kernels/imgproc/warp_cpu.h"
 #include "dali/kernels/imgproc/warp_gpu.h"
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/displacement/warp_param_provider.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/displacement/warp_param_provider.h"
```
Comment (klecki) [inline dali/operators/displacement/warp.h]: And here

File: dali/operators/displacement/warp_affine_params.h [hunk 14]
```diff
@@ -12,15 +12,15 @@
...
+#define DALI_OPERATORS_DISPLACEMENT_WARP_AFFINE_PARAMS_H_
 
 #include <vector>
 #include <sstream>
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/operator.h"
 #include "dali/kernels/imgproc/warp/affine.h"
 #include "dali/kernels/imgproc/warp/mapping_traits.h"
-#include "dali/pipeline/operators/displacement/warp_param_provider.h"
+#include "dali/operators/displacement/warp_param_provider.h"
```
Comment (klecki) [inline dali/operators/displacement/warp_affine_params.h]: And here

File: dali/operators/displacement/warpaffine.h [hunk 15]
```diff
@@ -13,13 +13,13 @@
...
+#ifndef DALI_OPERATORS_DISPLACEMENT_WARPAFFINE_H_
+#define DALI_OPERATORS_DISPLACEMENT_WARPAFFINE_H_
 
 #include <vector>
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/displacement/displacement_filter.h"
-#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/displacement/displacement_filter.h"
+#include "dali/pipeline/operator/common.h"
```
Comment (klecki) [inline dali/operators/displacement/warpaffine.h]: And here

File: dali/operators/expressions/expression_impl_factory_cpu.cc [hunk 16]
```diff
@@ -14,11 +14,11 @@
...
 
-#include "dali/pipeline/operators/expressions/expression_tree.h"
-#include "dali/pipeline/operators/expressions/arithmetic_meta.h"
+#include "dali/operators/expressions/expression_tree.h"
+#include "dali/operators/expressions/arithmetic_meta.h"
 #include "dali/core/static_switch.h"
-#include "dali/pipeline/operators/expressions/expression_impl_cpu.h"
-#include "dali/pipeline/operators/expressions/expression_impl_factory.h"
+#include "dali/operators/expressions/expression_impl_cpu.h"
+#include "dali/operators/expressions/expression_impl_factory.h"
```
Comment (klecki) [inline dali/operators/expressions/expression_impl_factory_cpu.cc]: Sorted it myself.

File: dali/operators/expressions/expression_tree.h [hunk 17]
```diff
@@ -23,7 +23,7 @@
 
 #include "dali/core/any.h"
 #include "dali/pipeline/data/types.h"
-#include "dali/pipeline/operators/expressions/arithmetic_meta.h"
+#include "dali/operators/expressions/arithmetic_meta.h"
```
Comment (klecki) [inline dali/operators/expressions/expression_tree.h]: And here

File: dali/operators/expressions/expression_tree_test.cc [hunk 18]
```diff
@@ -20,8 +20,8 @@
 
 #include "dali/core/static_switch.h"
 #include "dali/pipeline/data/types.h"
-#include "dali/pipeline/operators/expressions/arithmetic.h"
-#include "dali/pipeline/operators/expressions/arithmetic_meta.h"
+#include "dali/operators/expressions/arithmetic.h"
+#include "dali/operators/expressions/arithmetic_meta.h"
```
Comment (klecki) [inline dali/operators/expressions/expression_tree_test.cc]: And here

File: dali/operators/fused/crop_mirror_normalize.h [hunk 19]
```diff
@@ -26,9 +26,9 @@
 #include "dali/kernels/kernel_manager.h"
 #include "dali/kernels/scratch.h"
 #include "dali/kernels/slice/slice_flip_normalize_permute_common.h"
-#include "dali/pipeline/operators/common.h"
-#include "dali/pipeline/operators/crop/crop_attr.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/operators/crop/crop_attr.h"
+#include "dali/pipeline/operator/operator.h"
```
Comment (klecki) [inline dali/operators/fused/crop_mirror_normalize.h]: And here

File: dali/operators/reader/loader/loader.h [hunk 20]
```diff
@@ -28,9 +28,9 @@
 
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
-#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/operator/op_spec.h"
 #include "dali/pipeline/data/tensor.h"
-#include "dali/pipeline/operators/decoder/cache/image_cache_factory.h"
+#include "dali/operators/decoder/cache/image_cache_factory.h"
```
Comment (klecki) [inline dali/operators/reader/loader/loader.h]: And here

File: dali/operators/reader/loader/coco_loader.h [hunk 21]
```diff
@@ -12,16 +12,16 @@
...
 #include <algorithm>
 #include <memory>
 #include <string>
 #include <vector>
 #include <utility>
 
-#include "dali/pipeline/operators/reader/loader/file_loader.h"
+#include "dali/operators/reader/loader/file_loader.h"
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
```
Comment (klecki) [inline dali/operators/reader/loader/coco_loader.h]: And here

File: dali/operators/paste/paste.h [hunk 22]
```diff
@@ -12,18 +12,18 @@
...
 #include <utility>
 #include <vector>
 #include <random>
 
 #include "dali/core/common.h"
-#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operator/common.h"
 #include "dali/core/error_handling.h"
-#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operator/operator.h"
```
Comment (klecki) [inline dali/operators/paste/paste.h]: And here

File: dali/operators/optical_flow/optical_flow.h [hunk 23]
```diff
@@ -12,16 +12,16 @@
...
 #include <memory>
 #include <vector>
 #include "dali/pipeline/data/views.h"
 #include "dali/pipeline/data/backend.h"
-#include "dali/pipeline/operators/operator.h"
-#include "dali/pipeline/operators/optical_flow/optical_flow_adapter/optical_flow_stub.h"
-#include "dali/pipeline/operators/optical_flow/turing_of/optical_flow_turing.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/operators/optical_flow/optical_flow_adapter/optical_flow_stub.h"
+#include "dali/operators/optical_flow/turing_of/optical_flow_turing.h"
```
Comment (klecki) [inline dali/operators/optical_flow/optical_flow.h]: And here

File: dali/operators/optical_flow/optical_flow.cc [hunk 24]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 #include "dali/pipeline/data/views.h"
-#include "dali/pipeline/operators/optical_flow/optical_flow.h"
+#include "dali/operators/optical_flow/optical_flow.h"
```
Comment (klecki) [inline dali/operators/optical_flow/optical_flow.cc]: And here

File: dali/operators/geometric/shapes_test.cc [hunk 25]
```diff
@@ -14,7 +14,7 @@
 
 #include <gtest/gtest.h>
 #include <random>
-#include "dali/pipeline/operators/geometric/shapes.h"
+#include "dali/operators/geometric/shapes.h"
```
Comment (klecki) [inline dali/operators/geometric/shapes_test.cc]: And here

File: dali/operators/python_function/CMakeLists.txt [hunk 26]
```diff
@@ -14,12 +14,13 @@
...
 add_library(${python_function_lib} SHARED ${SOURCES})
-target_link_libraries(${python_function_lib} ${dali_lib})
-target_include_directories(${python_function_lib}
-        PRIVATE ${PYBIND11_INCLUDE_DIR}  # from project CMakeLists.txt
-        PRIVATE ${pybind11_INCLUDE_DIR}  # from pybind11Config
-        PRIVATE ${PYTHON_INCLUDE_DIRS})
+target_link_libraries(${python_function_lib} PUBLIC ${dali_lib})
+target_include_directories(${python_function_lib} PRIVATE
+        ${PYBIND11_INCLUDE_DIR}  # from project CMakeLists.txt
+        ${pybind11_INCLUDE_DIR}  # from pybind11Config
```
Comment (JanuszL) [inline dali/operators/python_function/CMakeLists.txt]: Why it cannot be private any longer?

Comment (jantonguirao) [reply dali/operators/python_function/CMakeLists.txt]: It is private, you don't need to repeat the keyword

Comment (JanuszL) [reply dali/operators/python_function/CMakeLists.txt]: OK

**Comment by jantonguirao** [comment]
Closing in favor of #1384

**Pull request closed without merging**